nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -1051,29 +1063,102 @@ impl<'a> Planner<'a> {
+        // Find out of date zones, as defined by zones whose image source does
+        // not match what it should be based on our current target release.
+        let target_release = self.input.tuf_repo().description();
+        let mut out_of_date_zones = sleds
             .into_iter()
             .flat_map(|sled_id| {
-                let blueprint = &self.blueprint;
-                blueprint
+                let log = &self.log;
+                self.blueprint
                     .current_sled_zones(
                         sled_id,
                         BlueprintZoneDisposition::is_in_service,
                     )
                     .filter_map(move |zone| {
-                        (zone.image_source
-                            != blueprint
-                                .zone_image_source(zone.zone_type.kind()))
-                        .then(|| (sled_id, zone.clone()))
+                        let desired_image_source = match target_release
+                            .zone_image_source(zone.zone_type.kind())
+                        {
+                            Ok(source) => source,
+                            Err(err) => {
+                                // If we can't tell whether a zone is out of
+                                // date, assume it isn't.
+                                warn!(
+                                    log,
+                                    "cannot determine whether zone is \
+                                     out of date";
+                                    "zone" => ?zone,
+                                    InlineErrorChain::new(&err),
+                                );
+                                return None;
+                            }
+                        };
+                        if zone.image_source != desired_image_source {
+                            Some((sled_id, zone, desired_image_source))
+                        } else {
+                            None
+                        }
                     })
             })
-            .collect::<Vec<(SledUuid, BlueprintZoneConfig)>>();
-        if let Some((sled_id, zone)) = out_of_date_zones.first() {
-            return self.update_or_expunge_zone(*sled_id, zone);
+            .peekable();
+
+        // Before we filter out zones that can't be updated, do we have any out
+        // of date zones at all? We need this to explain why we didn't update
+        // any zones below, if we don't.
+        let have_out_of_date_zones = out_of_date_zones.peek().is_some();
+
+        // Of the out-of-date zones, filter out zones that can't be updated yet,
+        // either because they're not ready or because it wouldn't be safe to
+        // bounce them.
+        let mut updateable_zones =
+            out_of_date_zones.filter(|(_sled_id, zone, _new_image_source)| {
+                let kind = zone.zone_type.kind();
+                if !self.can_zone_be_shut_down_safely(kind) {
```

Comment:
This separation makes sense to me. I was a little worried with the old format, where adding the check for "live_nodes == 5" because that could prevent us from adding a CRDB node when "live_nodes == 4" (which would be bad). 

But this structure should let us still make that check, and we only need to validate it in the "update-in-place" case, not necessarily in the "add-new-zone" case.